File: backend/src/cuanto_cuesta/domain/rates.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

from cuanto_cuesta.domain.errors import CurrencyMismatchError
from cuanto_cuesta.domain.money import Currency, Money, div, mul


@dataclass(frozen=True, slots=True)
class Rate:
    """Price of one unit of ``base`` expressed in ``quote``.

    ``bid`` is what you get when selling ``base``; ``ask`` is what you pay
    when buying it. The side is therefore implied by the direction of the
    conversion and never configured by hand.
    """

    key: str
    base: Currency
    quote: Currency
    bid: Decimal
    ask: Decimal

    def __post_init__(self) -> None:
        if self.base is self.quote:
            raise ValueError(f"Rate {self.key!r}: base and quote must differ")
        for name, price in (("bid", self.bid), ("ask", self.ask)):
            if not isinstance(price, Decimal) or not price.is_finite() or price <= 0:
                raise ValueError(f"Rate {self.key!r}: {name} must be a positive Decimal")

    def price_for(self, source: Currency, target: Currency) -> Decimal:
        """Price applied when converting from ``source`` to ``target``."""
        if source is self.base and target is self.quote:
            return self.bid
        if source is self.quote and target is self.base:
            return self.ask
        raise CurrencyMismatchError(
            f"Rate {self.key!r} ({self.base}/{self.quote}) cannot convert {source} to {target}"
        )

    def convert(self, amount: Money, target: Currency) -> Money:
        """Convert ``amount`` into ``target``, rounded down to the minor unit."""
        price = self.price_for(amount.currency, target)
        if amount.currency is self.base:
            converted = mul(amount.amount, price)
        else:
            converted = div(amount.amount, price)
        return Money(converted, target).rounded_down()

File: frontend/src/calculator/rates.ts
import {
  CurrencyMismatchError,
  Decimal,
  money,
  roundedDown,
  type Big,
  type Currency,
  type Money,
} from "./money.ts";

/**
 * Price of one unit of `base` in `quote`, as the person types it for their own operation (for
 * example ARS per USDT when selling USDT). Converting `base` to `quote` multiplies by it;
 * converting `quote` to `base` divides.
 */
export interface Rate {
  readonly key: string;
  readonly base: Currency;
  readonly quote: Currency;
  readonly price: Big;
}

export class InvalidRateError extends Error {
  override name = "InvalidRateError";
}

export function rate(key: string, base: Currency, quote: Currency, price: string | Big): Rate {
  const value = new Decimal(price);
  if (base === quote) {
    throw new InvalidRateError(`Rate ${key}: base and quote must differ`);
  }
  if (value.lte(0)) {
    throw new InvalidRateError(`Rate ${key}: price must be positive`);
  }
  return { key, base, quote, price: value };
}

/** Convert `amount` into `target`, rounded down to the minor unit. */
export function convert(rate: Rate, amount: Money, target: Currency): Money {
  if (amount.currency === rate.base && target === rate.quote) {
    return roundedDown(money(amount.amount.times(rate.price), target));
  }
  if (amount.currency === rate.quote && target === rate.base) {
    return roundedDown(money(amount.amount.div(rate.price), target));
  }
  throw new CurrencyMismatchError(
    `Rate ${rate.key} (${rate.base}/${rate.quote}) cannot convert ${amount.currency} to ${target}`,
  );
}
